### ISSUE TITLE
Error while making changes to Unity scenes in Ubuntu

### ISSUE THREAD

**Issue by sashi2295** [allenai/ai2thor#520]
I am trying to modify the scenes by adding objects to them. The README gives some information on how to do this using the Unity Editor. It states that the Linux Editor is in beta and that we need to install the Editor in OSX, but I was able to install the given version of the Unity Editor in Ubuntu and I was also able to open a scene and add an object to the scene.

I then saved the scene, closed the editor and ran the commands as specified:
`pip install invoke`
`invoke local-build`

and got the following error output:
```
Traceback (most recent call last):
  File "/home/sashi/software/miniconda3/envs/ai2-thor/bin/invoke", line 10, in <module>
    sys.exit(program.run())
  File "/home/sashi/software/miniconda3/envs/ai2-thor/lib/python3.8/site-packages/invoke/program.py", line 384, in run
    self.execute()
  File "/home/sashi/software/miniconda3/envs/ai2-thor/lib/python3.8/site-packages/invoke/program.py", line 566, in execute
    executor.execute(*self.tasks)
  File "/home/sashi/software/miniconda3/envs/ai2-thor/lib/python3.8/site-packages/invoke/executor.py", line 129, in execute
    result = call.task(*args, **call.kwargs)
  File "/home/sashi/software/miniconda3/envs/ai2-thor/lib/python3.8/site-packages/invoke/tasks.py", line 127, in __call__
    result = self.body(*args, **kwargs)
  File "/home/sashi/everything/ai2thor/tasks.py", line 345, in local_build
    if _build("unity", arch, "builds", build_name, env=env):
  File "/home/sashi/everything/ai2thor/tasks.py", line 95, in _build
    if os.path.exists(standalone_path):
UnboundLocalError: local variable 'standalone_path' referenced before assignment
```

I went through the code in `tasks.py` and it looks like the [code](https://github.com/allenai/ai2thor/blob/e9c33832a68f88c8be191682d09d68fb44902ad7/tasks.py#L83) is written only for Windows and Mac:

```python
def _build(unity_path, arch, build_dir, build_name, env={}):
    project_path = os.path.join(os.getcwd(), unity_path)
    if sys.platform.startswith('darwin'):
        unity_hub_path = "/Applications/Unity/Hub/Editor/{}/Unity.app/Contents/MacOS/Unity".format(
            UNITY_VERSION
        )
        standalone_path = "/Applications/Unity-{}/Unity.app/Contents/MacOS/Unity".format(UNITY_VERSION)
    elif 'win' in sys.platform:
        unity_hub_path = "C:/PROGRA~1/Unity/Hub/Editor/{}/Editor/Unity.exe".format(UNITY_VERSION)
        # TODO: Verify windows unity standalone path
        standalone_path = "C:/PROGRA~1/{}/Editor/Unity.exe".format(UNITY_VERSION)

    if os.path.exists(standalone_path):
        unity_path = standalone_path
    else:
        unity_path = unity_hub_path
    command = (
        "%s -quit -batchmode -logFile %s.log -projectpath %s -executeMethod Build.%s"
        % (unity_path, build_name, project_path, arch)
    )
```

Is there a fix that allows me to edit the scenes in Ubuntu?

**Comment by ekolve** [comment]
You will need to find the path to the Unity executable and create a block that checks the platform for 'linux' (similar to the mac/win blocks).  The issue I believe you are going to run in to is that the Unity Editor for Linux does not support using the Mono scripting backend which is what is used currently for OSX and Linux.  IL2CPP removes some .net features we currently use, as well we have yet to test IL2CPP builds to ensure they behave identically to the Mono builds.

**Comment by sashi2295** [comment]
Thank you for the super quick response!

I followed the steps that you have written and could get past the previous error. Now, when I do `invoke local build`, I get the following output:

```
/home/sashi/everything/ai2thor/tasks.py:83: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  project_version = yaml.load(pf.read())
CreateDirectory '' failed:  (current dir: )
CreateDirectory '' failed:  (current dir: )
Exited with code 0
Build Successful
```

Although it says `Build Successful`, there is no build beneath the directory `unity/builds/thor-local-OSXIntel64.app` as mentioned in the README. I am guessing this has something to do with the `CreateDirectory '' failed:  (current dir: )` output. I could isolate this output to this piece of [code](https://github.com/allenai/ai2thor/blob/7e5471156c3c6c65494b05b2b77580b742a5767c/tasks.py#L111) in `tasks.py`:

```python
def _build(unity_path, arch, build_dir, build_name, env={}):
    import yaml
    ...
    command = (
        "%s -quit -batchmode -logFile %s.log -projectpath %s -executeMethod Build.%s"
        % (unity_path, build_name, project_path, arch)
    )
    target_path = os.path.join(build_dir, build_name)

    full_env = os.environ.copy()
    full_env.update(env)
    full_env["UNITY_BUILD_NAME"] = target_path
    result_code = subprocess.check_call(command, shell=True, env=full_env)
```

The `command` in the above code comes down to:
```
/home/sashi/Unity/Hub/Editor/2019.4.2f1/Editor/Unity -quit -batchmode -logFile thor-local-OSXIntel64.log -projectpath /home/sashi/everything/ai2thor/unity -executeMethod Build.OSXIntel64
```

Executing this directly in the terminal gives me:
```
CreateDirectory '' failed:  (current dir: )
CreateDirectory '' failed:  (current dir: )

Aborting batchmode due to failure:
executeMethod method Build.OSXIntel64 threw exception.
```

Maybe more changes are needed to adapt to the Linux platform?

**Comment by ekolve** [comment]
I think i may know the issue.  I installed Unity 2019.4.13f1 and opened the project with that version.  Can you try installing that version with Unity Hub and open the project with that in the Editor (this will change the ProjectVersion in the ProjectSettings), then close the editor and run invoke local-build.  You can pass --arch=Linux64 to local-build to build the Linux version.

**Comment by sashi2295** [comment]
The build is successful now! I can start a controller using the new build and visualize my modified scene as well! Thanks a lot for your help!

I can now move the agent in this new scene, but it looks like the new object that has been added isn't loaded into the metadata. In my case, I added a `DeskLamp` to `FloorPlan201` and placed it on the `Floor`. Now, when I use `controller.interact()` and move the agent close to this `DeskLamp`, it doesn't get listed in the `Visible Objects` list.

I also checked the length of the objects list in the metadata from the original scene and from the modified scene. In both scenes, the number of objects is the same.

Is this expected behavior or is this new object supposed to be in the metadata as well?
